--- a/net/minecraft/client/renderer/model/ModelBakery.java
+++ b/net/minecraft/client/renderer/model/ModelBakery.java
@@ -247,7 +247,7 @@
                             try
                             {
                                 flag = true;
-                                lvt_9_4_ = this.resourceManager.getAllResources(resourcelocation1).stream().map((p_209591_1_) ->
+                                lvt_9_4_ = (List)this.resourceManager.getAllResources(resourcelocation1).stream().map((p_209591_1_) ->
                                 {
 
                                     try (InputStream inputstream = p_209591_1_.getInputStream())
