--- a/net/minecraft/command/impl/CloneCommand.java
+++ b/net/minecraft/command/impl/CloneCommand.java
@@ -1,5 +1,6 @@
 package net.minecraft.command.impl;
 
+import carpet.CarpetSettings;
 import com.google.common.collect.Lists;
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
@@ -107,9 +108,9 @@
         {
             int i = mutableboundingbox.getXSize() * mutableboundingbox.getYSize() * mutableboundingbox.getZSize();
 
-            if (i > 32768)
+            if (i > CarpetSettings.getInt("fillLimit")) // [CM] replaces 32768
             {
-                throw CLONE_TOO_BIG_EXCEPTION.create(32768, i);
+                throw CLONE_TOO_BIG_EXCEPTION.create(CarpetSettings.getInt("fillLimit"), i);
             }
             else
             {
@@ -170,12 +171,12 @@
                                 ((IInventory)tileentity1).clear();
                             }
 
-                            worldserver.setBlockState(blockpos3, Blocks.BARRIER.getDefaultState(), 2);
+                            worldserver.setBlockState(blockpos3, Blocks.BARRIER.getDefaultState(), 2 | (CarpetSettings.getBool("fillUpdates")?0:1024)); // [CM]
                         }
 
                         for (BlockPos blockpos4 : deque)
                         {
-                            worldserver.setBlockState(blockpos4, Blocks.AIR.getDefaultState(), 3);
+                            worldserver.setBlockState(blockpos4, Blocks.AIR.getDefaultState(), 3 | (CarpetSettings.getBool("fillUpdates")?0:1024)); // [CM]
                         }
                     }
 
@@ -194,14 +195,14 @@
                             ((IInventory)tileentity2).clear();
                         }
 
-                        worldserver.setBlockState(clonecommand$blockinfo.pos, Blocks.BARRIER.getDefaultState(), 2);
+                        worldserver.setBlockState(clonecommand$blockinfo.pos, Blocks.BARRIER.getDefaultState(), 2 | (CarpetSettings.getBool("fillUpdates")?0:1024)); // [CM]
                     }
 
                     int i1 = 0;
 
                     for (CloneCommand.BlockInfo clonecommand$blockinfo1 : list3)
                     {
-                        if (worldserver.setBlockState(clonecommand$blockinfo1.pos, clonecommand$blockinfo1.state, 2))
+                        if (worldserver.setBlockState(clonecommand$blockinfo1.pos, clonecommand$blockinfo1.state, 2 | (CarpetSettings.getBool("fillUpdates")?0:1024)))
                         {
                             ++i1;
                         }
@@ -220,8 +221,11 @@
                             tileentity3.markDirty();
                         }
 
-                        worldserver.setBlockState(clonecommand$blockinfo2.pos, clonecommand$blockinfo2.state, 2);
+                        worldserver.setBlockState(clonecommand$blockinfo2.pos, clonecommand$blockinfo2.state, 2 | (CarpetSettings.getBool("fillUpdates")?0:1024)); // [CM]
                     }
+                    /*carpet mod */
+                    if (CarpetSettings.getBool("fillUpdates"))
+                    {   /*carpet mod end EXTRA INDENTATION START*/
 
                     for (CloneCommand.BlockInfo clonecommand$blockinfo3 : list4)
                     {
@@ -230,6 +234,8 @@
 
                     worldserver.getPendingBlockTicks().copyTicks(mutableboundingbox, blockpos);
 
+                    } //carpet mod back extra indentation
+
                     if (i1 == 0)
                     {
                         throw FAILED_EXCEPTION.create();
