--- a/net/minecraft/command/impl/FillCommand.java
+++ b/net/minecraft/command/impl/FillCommand.java
@@ -1,5 +1,6 @@
 package net.minecraft.command.impl;
 
+import carpet.CarpetSettings;
 import com.google.common.collect.Lists;
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
@@ -70,9 +71,9 @@
     {
         int i = area.getXSize() * area.getYSize() * area.getZSize();
 
-        if (i > 32768)
+        if (i > CarpetSettings.fillLimit) // [CM] replaces 32768
         {
-            throw TOO_BIG_EXCEPTION.create(32768, i);
+            throw TOO_BIG_EXCEPTION.create(CarpetSettings.fillLimit, i);
         }
         else
         {
@@ -95,7 +96,7 @@
                             ((IInventory)tileentity).clear();
                         }
 
-                        if (blockstateinput.place(worldserver, blockpos, 2))
+                        if (blockstateinput.place(worldserver, blockpos, 2 | (CarpetSettings.fillUpdates?0:1024))) // [CM]
                         {
                             list.add(blockpos.toImmutable());
                             ++j;
@@ -103,12 +104,16 @@
                     }
                 }
             }
+            /*carpet mod */
+            if (CarpetSettings.fillUpdates)
+            { /*carpet mod end EXTRA INDENT*/
 
             for (BlockPos blockpos1 : list)
             {
                 Block block = worldserver.getBlockState(blockpos1).getBlock();
                 worldserver.notifyNeighbors(blockpos1, block);
             }
+            } //carpet mod back extra indentation
 
             if (j == 0)
             {
