--- a/net/minecraft/world/dimension/OverworldDimension.java
+++ b/net/minecraft/world/dimension/OverworldDimension.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.dimension;
 
+import carpet.CarpetSettings;
 import com.google.gson.JsonArray;
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
@@ -55,7 +56,12 @@
 
     public boolean canDropChunk(int x, int z)
     {
-        return !this.world.isSpawnChunk(x, z) && super.canDropChunk(x, z);
+        // CM Changed to support disabling spawn chunks
+        // return !this.world.isSpawnChunk(x, z) && super.canDropChunk(x, z);
+        if (this.world.isSpawnChunk(x, z))
+            return CarpetSettings.disableSpawnChunks && super.canDropChunk(x, z);
+
+        return super.canDropChunk(x, z);
     }
 
     protected void init()
