--- a/net/minecraft/world/gen/feature/structure/DesertPyramidStructure.java
+++ b/net/minecraft/world/gen/feature/structure/DesertPyramidStructure.java
@@ -1,5 +1,7 @@
 package net.minecraft.world.gen.feature.structure;
 
+import com.google.common.collect.Lists;
+import net.minecraft.entity.EntityType;
 import net.minecraft.init.Biomes;
 import net.minecraft.util.SharedSeedRandom;
 import net.minecraft.util.math.BlockPos;
@@ -7,8 +9,12 @@
 import net.minecraft.world.biome.Biome;
 import net.minecraft.world.gen.IChunkGenerator;
 
+import java.util.List;
+
 public class DesertPyramidStructure extends ScatteredStructure<DesertPyramidConfig>
 {
+    private static final List<Biome.SpawnListEntry> spawnList = Lists.newArrayList(new Biome.SpawnListEntry(EntityType.HUSK, 1, 1, 1));
+
     protected String getStructureName()
     {
         return "Desert_Pyramid";
@@ -25,6 +31,12 @@
         return new DesertPyramidStructure.Start(worldIn, random, x, z, biome);
     }
 
+    @Override
+    public List<Biome.SpawnListEntry> getSpawnList()
+    {
+        return spawnList;
+    }
+
     protected int getSeedModifier()
     {
         return 14357617;
