--- a/net/minecraft/world/gen/feature/structure/EndCityStructure.java
+++ b/net/minecraft/world/gen/feature/structure/EndCityStructure.java
@@ -1,7 +1,14 @@
 package net.minecraft.world.gen.feature.structure;
 
+import java.util.List;
 import java.util.Random;
+
+import carpet.CarpetSettings;
+import com.google.common.collect.Lists;
+import net.minecraft.entity.EntityType;
+import net.minecraft.entity.monster.EntityShulker;
 import net.minecraft.init.Biomes;
+import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.util.Rotation;
 import net.minecraft.util.SharedSeedRandom;
 import net.minecraft.util.math.BlockPos;
@@ -16,6 +23,8 @@
 
 public class EndCityStructure extends Structure<EndCityConfig>
 {
+    private static final List<Biome.SpawnListEntry> spawnList = Lists.newArrayList(new Biome.SpawnListEntry(EntityType.SHULKER, 10, 4, 4));
+
     protected ChunkPos getStartPositionForPosition(IChunkGenerator<?> chunkGenerator, Random random, int x, int z, int spacingOffsetsX, int spacingOffsetsZ)
     {
         int i = chunkGenerator.getSettings().getEndCityDistance();
@@ -79,6 +88,12 @@
         return 9;
     }
 
+    @Override
+    public List<Biome.SpawnListEntry> getSpawnList()
+    {
+        return spawnList;
+    }
+
     private static int getYPosForStructure(int chunkX, int chunkY, IChunkGenerator<?> generatorIn)
     {
         Random random = new Random((long)(chunkX + chunkY * 10387313));
@@ -140,5 +155,28 @@
             {
                 return this.isSizeable;
             }
+
+            // [CM] Fix EndCity saving
+            @Override
+            public void writeAdditional(NBTTagCompound tagCompound)
+            {
+                super.writeAdditional(tagCompound);
+                if (CarpetSettings.getBool("endCitySavingFix"))
+                {
+                    // Called the tag Valid because it is called Valid in the VillageStructure too
+                    // and it is used to invalidate the Structure
+                    tagCompound.setBoolean("Valid", this.isSizeable);
+                }
+            }
+
+            @Override
+            public void readAdditional(NBTTagCompound tagCompound)
+            {
+                super.readAdditional(tagCompound);
+                if (CarpetSettings.getBool("endCitySavingFix") && tagCompound.hasKey("Valid"))
+                {
+                    isSizeable = tagCompound.getBoolean("Valid");
+                }
+            }
         }
 }
