--- a/net/minecraft/world/gen/feature/structure/StructurePiece.java
+++ b/net/minecraft/world/gen/feature/structure/StructurePiece.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.gen.feature.structure;
 
+import carpet.CarpetSettings;
 import com.google.common.collect.ImmutableSet;
 import java.util.List;
 import java.util.Random;
@@ -277,7 +278,7 @@
         }
         else
         {
-            return j < worldIn.getHeight(Heightmap.Type.OCEAN_FLOOR_WG, i, k);
+            return j < worldIn.getHeight(CarpetSettings.skipGenerationChecks?Heightmap.Type.OCEAN_FLOOR:Heightmap.Type.OCEAN_FLOOR_WG, i, k);
         }
     }
 
