--- a/net/minecraft/entity/passive/AbstractFish.java
+++ b/net/minecraft/entity/passive/AbstractFish.java
@@ -33,6 +33,11 @@
 {
     private static final DataParameter<Boolean> FROM_BUCKET = EntityDataManager.createKey(AbstractFish.class, DataSerializers.BOOLEAN);
 
+    //CM
+    @Override
+    public String cm_name() { return "Fish"; }
+
+
     public AbstractFish(EntityType<?> type, World p_i48855_2_)
     {
         super(type, p_i48855_2_);
