--- a/net/minecraft/entity/passive/EntityAmbientCreature.java
+++ b/net/minecraft/entity/passive/EntityAmbientCreature.java
@@ -7,6 +7,10 @@
 
 public abstract class EntityAmbientCreature extends EntityLiving implements IAnimal
 {
+    //CM
+    @Override
+    public String cm_name() { return "Bat"; }
+
     protected EntityAmbientCreature(EntityType<?> type, World p_i48570_2_)
     {
         super(type, p_i48570_2_);
