--- a/net/minecraft/tileentity/TileEntityFurnace.java
+++ b/net/minecraft/tileentity/TileEntityFurnace.java
@@ -53,6 +53,12 @@
     private ITextComponent furnaceCustomName;
     private final Map<ResourceLocation, Integer> recipeUseCounts = Maps.newHashMap();
 
+
+    //CM
+    @Override
+    public String cm_name() { return "Furnace"; }
+
+
     private static void setBurnTime(Map<Item, Integer> map, Tag<Item> tagIn, int time)
     {
         for (Item item : tagIn.getAllElements())
