--- a/net/minecraft/entity/EntityLiving.java
+++ b/net/minecraft/entity/EntityLiving.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity;
 
+import carpet.logging.logHelpers.DamageReporter;
 import com.google.common.collect.Maps;
 import java.util.Arrays;
 import java.util.Map;
@@ -1447,10 +1448,16 @@
         float f = (float)this.getAttribute(SharedMonsterAttributes.ATTACK_DAMAGE).getValue();
         int i = 0;
 
+        // CM
+        DamageReporter.register_damage_attacker( entityIn, this, f);
+        float ff = f;
+        // end CM
+
         if (entityIn instanceof EntityLivingBase)
         {
             f += EnchantmentHelper.getModifierForCreature(this.getHeldItemMainhand(), ((EntityLivingBase)entityIn).getCreatureAttribute());
             i += EnchantmentHelper.getKnockbackModifier(this);
+            DamageReporter.modify_damage((EntityLivingBase)entityIn, DamageSource.causeMobDamage(this), ff, f, "attacker enchants"); // CM
         }
 
         boolean flag = entityIn.attackEntityFrom(DamageSource.causeMobDamage(this), f);
@@ -1522,4 +1529,28 @@
             this.motionY += (double)0.3F;
         }
     }
+
+    // Added optimized despawn mobs by Luflosi
+    public boolean willImmediatelyDespawn()
+    {
+        if (!this.canDespawn() || this.persistenceRequired)
+        {
+            return false;
+        }
+        boolean playerInDimension = false;
+        for (int i = 0; i < this.world.playerEntities.size(); i++)
+        {
+            EntityPlayer entityplayer = this.world.playerEntities.get(i);
+            if (!entityplayer.isSpectator())
+            {
+                playerInDimension = true;
+                double distanceSq = entityplayer.getDistanceSq(this.posX, this.posY, this.posZ);
+                if (distanceSq <= 16384.0D)
+                {
+                    return false;
+                }
+            }
+        }
+        return playerInDimension;
+    }
 }
