--- a/net/minecraft/entity/EntityLiving.java
+++ b/net/minecraft/entity/EntityLiving.java
@@ -1,5 +1,7 @@
 package net.minecraft.entity;
 
+import carpet.CarpetSettings;
+import carpet.logging.logHelpers.DamageReporter;
 import com.google.common.collect.Maps;
 import java.util.Arrays;
 import java.util.Map;
@@ -836,7 +838,7 @@
 
     public boolean isNotColliding(IWorldReaderBase worldIn)
     {
-        return !worldIn.containsAnyLiquid(this.getBoundingBox()) && worldIn.isCollisionBoxesEmpty(this, this.getBoundingBox()) && worldIn.checkNoEntityCollision(this, this.getBoundingBox());
+        return !worldIn.containsAnyLiquid(this.getBoundingBox()) && worldIn.isCollisionBoxesEmpty(this, this.getBoundingBox()) && (CarpetSettings.n_mobSpawningAlgorithm == 112 || worldIn.checkNoEntityCollision(this, this.getBoundingBox()));
     }
 
     @OnlyIn(Dist.CLIENT)
@@ -1447,10 +1449,16 @@
         float f = (float)this.getAttribute(SharedMonsterAttributes.ATTACK_DAMAGE).getValue();
         int i = 0;
 
+        // CM
+        DamageReporter.register_damage_attacker( entityIn, this, f);
+        float ff = f;
+        // end CM
+
         if (entityIn instanceof EntityLivingBase)
         {
             f += EnchantmentHelper.getModifierForCreature(this.getHeldItemMainhand(), ((EntityLivingBase)entityIn).getCreatureAttribute());
             i += EnchantmentHelper.getKnockbackModifier(this);
+            DamageReporter.modify_damage((EntityLivingBase)entityIn, DamageSource.causeMobDamage(this), ff, f, "attacker enchants"); // CM
         }
 
         boolean flag = entityIn.attackEntityFrom(DamageSource.causeMobDamage(this), f);
