--- a/net/minecraft/pathfinding/PathNavigate.java
+++ b/net/minecraft/pathfinding/PathNavigate.java
@@ -1,6 +1,10 @@
 package net.minecraft.pathfinding;
 
 import javax.annotation.Nullable;
+
+import carpet.CarpetSettings;
+import carpet.logging.LoggerRegistry;
+import carpet.logging.logHelpers.PathfindingVisualizer;
 import net.minecraft.block.Block;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.Entity;
@@ -105,6 +109,11 @@
         }
         else
         {
+            // CM pathfinding logger
+            long start = 0;
+            if (LoggerRegistry.__pathfinding)
+                start = System.nanoTime();
+            //
             this.targetPos = pos;
             float f = this.getPathSearchRange();
             this.world.profiler.startSection("pathfind");
@@ -113,6 +122,15 @@
             IBlockReader iblockreader = new Region(this.world, blockpos.add(-i, -i, -i), blockpos.add(i, i, i), 0);
             Path path = this.pathFinder.findPath(iblockreader, this.entity, this.targetPos, f);
             this.world.profiler.endSection();
+            //CM
+            if (LoggerRegistry.__pathfinding)
+            {
+                long end = System.nanoTime();
+                float duration = (1.0F*((end - start)/1000))/1000;
+                PathfindingVisualizer.slowPath(entity, new Vec3d(pos), duration, path != null);
+
+            }
+            ///
             return path;
         }
     }
@@ -134,6 +152,11 @@
             }
             else
             {
+                // CM pathfinding logger
+                long start = 0;
+                if (LoggerRegistry.__pathfinding)
+                    start = System.nanoTime();
+                //
                 this.targetPos = blockpos;
                 float f = this.getPathSearchRange();
                 this.world.profiler.startSection("pathfind");
@@ -142,6 +165,15 @@
                 IBlockReader iblockreader = new Region(this.world, blockpos1.add(-i, -i, -i), blockpos1.add(i, i, i), 0);
                 Path path = this.pathFinder.findPath(iblockreader, this.entity, entityIn, f);
                 this.world.profiler.endSection();
+                //CM
+                if (LoggerRegistry.__pathfinding)
+                {
+                    long end = System.nanoTime();
+                    float duration = (1.0F*((end - start)/1000))/1000;
+                    PathfindingVisualizer.slowPath(entity, entity.getPositionVector(), duration, path != null);
+
+                }
+                ///
                 return path;
             }
         }
