--- a/net/minecraft/entity/ai/EntityAIFindEntityNearestPlayer.java
+++ b/net/minecraft/entity/ai/EntityAIFindEntityNearestPlayer.java
@@ -18,7 +18,8 @@
 public class EntityAIFindEntityNearestPlayer extends EntityAIBase
 {
     private static final Logger LOGGER = LogManager.getLogger();
-    private final EntityLiving entityLiving;
+    private final EntityLiving entityLiving; // CM so I can access
+    protected EntityLiving getEntityLiving() { return entityLiving; }
     private final Predicate<Entity> predicate;
     private final EntityAINearestAttackableTarget.Sorter sorter;
     private EntityLivingBase entityTarget;
