--- a/net/minecraft/entity/ai/EntityMoveHelper.java
+++ b/net/minecraft/entity/ai/EntityMoveHelper.java
@@ -17,6 +17,8 @@
     protected float moveForward;
     protected float moveStrafe;
     protected EntityMoveHelper.Action action = EntityMoveHelper.Action.WAIT;
+    //CM need public access for ghast helper
+    public void setAction(Action action) { this.action = action; }
 
     public EntityMoveHelper(EntityLiving entitylivingIn)
     {
