--- a/net/minecraft/client/multiplayer/WorldClient.java
+++ b/net/minecraft/client/multiplayer/WorldClient.java
@@ -1,5 +1,6 @@
 package net.minecraft.client.multiplayer;
 
+import carpet.CarpetSettings;
 import com.google.common.collect.Sets;
 import java.util.Random;
 import java.util.Set;
@@ -129,7 +130,15 @@
     protected void refreshVisibleChunks()
     {
         this.visibleChunks.clear();
-        int i = this.mc.gameSettings.renderDistanceChunks;
+
+        //[CM] viewDistance override
+        int i;
+        if (CarpetSettings.clientViewDistance == 0)
+            i = this.mc.gameSettings.renderDistanceChunks;
+        else {
+            i = CarpetSettings.clientViewDistance;
+        }
+
         this.profiler.startSection("buildList");
         int j = MathHelper.floor(this.mc.player.posX / 16.0D);
         int k = MathHelper.floor(this.mc.player.posZ / 16.0D);
