--- a/net/minecraft/block/state/BlockPistonStructureHelper.java
+++ b/net/minecraft/block/state/BlockPistonStructureHelper.java
@@ -1,5 +1,6 @@
 package net.minecraft.block.state;
 
+import carpet.CarpetSettings;
 import com.google.common.collect.Lists;
 import java.util.List;
 import net.minecraft.block.Block;
@@ -103,7 +104,7 @@
         {
             int i = 1;
 
-            if (i + this.toMove.size() > 12)
+            if (i + this.toMove.size() > CarpetSettings.n_pushLimit)
             {
                 return false;
             }
@@ -122,7 +123,7 @@
 
                     ++i;
 
-                    if (i + this.toMove.size() > 12)
+                    if (i + this.toMove.size() > CarpetSettings.n_pushLimit)
                     {
                         return false;
                     }
@@ -178,7 +179,7 @@
                         return true;
                     }
 
-                    if (this.toMove.size() >= 12)
+                    if (this.toMove.size() >= CarpetSettings.n_pushLimit)
                     {
                         return false;
                     }
