--- a/net/minecraft/entity/item/EntityItem.java
+++ b/net/minecraft/entity/item/EntityItem.java
@@ -37,6 +37,11 @@
     private UUID owner;
     public float hoverStart = (float)(Math.random() * Math.PI * 2.0D);
 
+    //CM
+    @Override
+    public String cm_name() { return "Item"; }
+
+
     public EntityItem(World worldIn)
     {
         super(EntityType.ITEM, worldIn);
@@ -227,8 +232,22 @@
                     }
                     else if (itemstack1.getCount() + itemstack.getCount() > itemstack1.getMaxStackSize())
                     {
+                        // Add check for stackableShulkerBoxes without NBT on the ground Xcom/Salandora
+                        if (itemstack1.isGroundStackable() && itemstack.isGroundStackable())
+                        {
+                            itemstack1.grow(itemstack.getCount());
+                            other.pickupDelay = Math.max(other.pickupDelay, this.pickupDelay);
+                            other.age = Math.min(other.age, this.age);
+                            other.setItem(itemstack1);
+                            this.remove();
+                            return true;
+                        }
                         return false;
                     }
+                    else if (!itemstack1.isStackable() && !itemstack.isStackable())
+                    {
+                        return false;
+                    }
                     else
                     {
                         itemstack1.grow(itemstack.getCount());
