--- a/net/minecraft/world/gen/ChunkGeneratorEnd.java
+++ b/net/minecraft/world/gen/ChunkGeneratorEnd.java
@@ -1,6 +1,8 @@
 package net.minecraft.world.gen;
 
 import java.util.List;
+
+import carpet.CarpetSettings;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.EnumCreatureType;
 import net.minecraft.init.Blocks;
@@ -14,6 +16,7 @@
 import net.minecraft.world.biome.provider.BiomeProvider;
 import net.minecraft.world.chunk.ChunkStatus;
 import net.minecraft.world.chunk.IChunk;
+import net.minecraft.world.gen.feature.Feature;
 
 public class ChunkGeneratorEnd extends AbstractChunkGenerator<EndGenSettings>
 {
@@ -203,6 +206,14 @@
 
     public List<Biome.SpawnListEntry> getPossibleCreatures(EnumCreatureType creatureType, BlockPos pos)
     {
+        //[CM] Add shulkers to the list of spawnable Creatures
+        if (CarpetSettings.b_shulkerSpawningInEndCities && EnumCreatureType.MONSTER == creatureType)
+        {
+            if (Feature.END_CITY.isPositionInsideStructure( this.world, pos))
+            {
+                return Feature.END_CITY.getSpawnList();
+            }
+        }
         return this.world.getBiome(pos).getSpawns(creatureType);
     }
 
