--- a/net/minecraft/world/WorldServer.java
+++ b/net/minecraft/world/WorldServer.java
@@ -1,5 +1,8 @@
 package net.minecraft.world;
 
+import carpet.CarpetSettings;
+import carpet.helpers.TickSpeed;
+import carpet.utils.CarpetProfiler;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.util.concurrent.ListenableFuture;
@@ -191,13 +194,25 @@
             this.wakeAllPlayers();
         }
 
+        //CM profiler
+        int did = this.dimension.getType().getId();
+        String world_name = (did==0)?"Overworld":((did<0?"The Nether":"The End"));
+        // CM end
+
+        if (TickSpeed.process_entities)
+        { // [CM] extra indent to skip processing of entities
         this.profiler.startSection("spawner");
+        CarpetProfiler.start_section(world_name, "Spawning");
 
-        if (this.getGameRules().getBoolean("doMobSpawning") && this.worldInfo.getTerrainType() != WorldType.DEBUG_ALL_BLOCK_STATES)
+
+            if (this.getGameRules().getBoolean("doMobSpawning") && this.worldInfo.getTerrainType() != WorldType.DEBUG_ALL_BLOCK_STATES)
         {
             this.entitySpawner.findChunksForSpawning(this, this.spawnHostileMobs, this.spawnPeacefulMobs, this.worldInfo.getGameTime() % 400L == 0L);
             this.getChunkProvider().spawnMobs(this, this.spawnHostileMobs, this.spawnPeacefulMobs);
         }
+        CarpetProfiler.end_current_section();
+        }
+        // [CM] end extra indent
 
         this.profiler.endStartSection("chunkSource");
         this.chunkProvider.tick(p_72835_1_);
@@ -208,6 +223,8 @@
             this.setSkylightSubtracted(j);
         }
 
+        if (TickSpeed.process_entities)
+        { // CM extra indent to skip processing of entities
         this.worldInfo.setWorldTotalTime(this.worldInfo.getGameTime() + 1L);
 
         if (this.getGameRules().getBoolean("doDaylightCycle"))
@@ -216,16 +233,27 @@
         }
 
         this.profiler.endStartSection("tickPending");
+        CarpetProfiler.start_section(world_name, "Blocks");
         this.tickPending();
+        CarpetProfiler.end_current_section();
+        } // end extra indent
         this.profiler.endStartSection("tickBlocks");
+        CarpetProfiler.start_section(world_name, "Blocks");
         this.tickBlocks();
+        CarpetProfiler.end_current_section();
         this.profiler.endStartSection("chunkMap");
         this.playerChunkMap.tick();
+
+        if (TickSpeed.process_entities)
+        { // CM extra indent to skip processing of entities
         this.profiler.endStartSection("village");
+        CarpetProfiler.start_section(world_name, "Villages");
         this.villageCollection.tick();
         this.villageSiege.tick();
+        CarpetProfiler.end_current_section();
         this.profiler.endStartSection("portalForcer");
         this.worldTeleporter.removeStalePortalLocations(this.getGameTime());
+        }
         this.profiler.endSection();
         this.sendQueuedBlockEvents();
         this.insideTick = false;
@@ -269,8 +297,18 @@
                     ++j;
                 }
             }
-
-            this.allPlayersSleeping = j > 0 && j >= this.playerEntities.size() - i;
+            // [CM] Added variable int & Conditional statement -- SleepingThreshold
+            int sleepingThreshold = CarpetSettings.sleepingThreshold;
+            if (sleepingThreshold < 100)
+            {
+                // Custom Amount
+                this.allPlayersSleeping = j > 0;
+            }
+            else
+            {
+                // Vanilla Amount
+                this.allPlayersSleeping = j > 0 && j >= this.playerEntities.size() - i;
+            }
         }
     }
 
@@ -306,6 +344,28 @@
     {
         if (this.allPlayersSleeping && !this.isRemote)
         {
+            // [CM] Added variable int and conditional statement -- SleepingThreshold
+            int sleepingThreshold = CarpetSettings.sleepingThreshold;
+            if (sleepingThreshold < 100)
+            {
+                float numberOfPlayers = 0;
+                float sleepingPlayerCount = 0;
+
+                for (EntityPlayer entityplayer : this.playerEntities)
+                {
+                    if (!entityplayer.isSpectator())
+                    {
+                        numberOfPlayers++;
+                        if (entityplayer.isPlayerFullyAsleep())
+                        {
+                            sleepingPlayerCount++;
+                        }
+                    }
+                }
+                return sleepingThreshold <= (int) ((sleepingPlayerCount / numberOfPlayers) * 100);
+            }
+
+            // [CM] end
             for (EntityPlayer entityplayer : this.playerEntities)
             {
                 if (!entityplayer.isSpectator() && !entityplayer.isPlayerFullyAsleep())
@@ -407,6 +467,13 @@
                 chunk.enqueueRelightChecks();
                 this.profiler.endStartSection("tickChunk");
                 chunk.tick(false);
+                //[CM] tick freeze
+                if (!TickSpeed.process_entities)
+                { // skipping the rest of the block processing
+                    this.profiler.endSection();
+                    continue;
+                }
+
                 this.profiler.endStartSection("thunder");
 
                 if (flag && flag1 && this.rand.nextInt(100000) == 0)
@@ -499,7 +566,7 @@
         }
     }
 
-    protected BlockPos adjustPosToNearbyEntity(BlockPos pos)
+    public BlockPos adjustPosToNearbyEntity(BlockPos pos) // [CM] Changed access to public for summonNaturalLightning
     {
         BlockPos blockpos = this.getHeight(Heightmap.Type.MOTION_BLOCKING, pos);
         AxisAlignedBB axisalignedbb = (new AxisAlignedBB(blockpos, new BlockPos(blockpos.getX(), this.getHeight(), blockpos.getZ()))).grow(3.0D);
