--- a/net/minecraft/entity/boss/EntityDragon.java
+++ b/net/minecraft/entity/boss/EntityDragon.java
@@ -79,6 +79,11 @@
     private final int[] neighbors = new int[24];
     private final PathHeap pathFindQueue = new PathHeap();
 
+    //CM
+    @Override
+    public String cm_name() { return "Dragon"; }
+
+
     public EntityDragon(World worldIn)
     {
         super(EntityType.ENDER_DRAGON, worldIn);
