--- a/net/minecraft/entity/boss/EntityWither.java
+++ b/net/minecraft/entity/boss/EntityWither.java
@@ -69,6 +69,11 @@
         return p_210135_0_ instanceof EntityLivingBase && ((EntityLivingBase)p_210135_0_).getCreatureAttribute() != CreatureAttribute.UNDEAD && ((EntityLivingBase)p_210135_0_).attackable();
     };
 
+    //CM
+    @Override
+    public String cm_name() { return "Wither"; }
+
+
     public EntityWither(World worldIn)
     {
         super(EntityType.WITHER, worldIn);
