--- a/net/minecraft/server/integrated/IntegratedServer.java
+++ b/net/minecraft/server/integrated/IntegratedServer.java
@@ -38,6 +38,8 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.CarpetServer;
+
 @OnlyIn(Dist.CLIENT)
 public class IntegratedServer extends MinecraftServer
 {
@@ -103,6 +105,10 @@
         this.setKeyPair(CryptManager.generateKeyPair());
         this.loadAllWorlds(this.getFolderName(), this.getWorldName(), this.worldSettings.getSeed(), this.worldSettings.getTerrainType(), this.worldSettings.getGeneratorOptions());
         this.setMOTD(this.getServerOwner() + " - " + this.getWorld(DimensionType.OVERWORLD).getWorldInfo().getWorldName());
+        //CM init - all stuff loaded from the server, just before worlds loading
+        CarpetServer.onServerLoaded(this);
+        //CM start game hook
+        CarpetServer.onGameStarted();
         return true;
     }
 
