--- a/net/minecraft/server/integrated/IntegratedServer.java
+++ b/net/minecraft/server/integrated/IntegratedServer.java
@@ -1,5 +1,6 @@
 package net.minecraft.server.integrated;
 
+import carpet.CarpetSettings;
 import com.google.common.collect.Lists;
 import com.google.common.util.concurrent.Futures;
 import com.google.gson.JsonElement;
@@ -38,6 +39,8 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.CarpetServer;
+
 @OnlyIn(Dist.CLIENT)
 public class IntegratedServer extends MinecraftServer
 {
@@ -101,6 +104,8 @@
         this.setAllowFlight(true);
         LOGGER.info("Generating keypair");
         this.setKeyPair(CryptManager.generateKeyPair());
+        //CM init - all stuff loaded from the server, just before worlds loading
+        CarpetServer.onServerLoaded(this);
         this.loadAllWorlds(this.getFolderName(), this.getWorldName(), this.worldSettings.getSeed(), this.worldSettings.getTerrainType(), this.worldSettings.getGeneratorOptions());
         this.setMOTD(this.getServerOwner() + " - " + this.getWorld(DimensionType.OVERWORLD).getWorldInfo().getWorldName());
         return true;
