--- a/net/minecraft/block/BlockShulkerBox.java
+++ b/net/minecraft/block/BlockShulkerBox.java
@@ -2,6 +2,8 @@
 
 import java.util.List;
 import javax.annotation.Nullable;
+
+import carpet.CarpetSettings;
 import net.minecraft.block.material.EnumPushReaction;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.IBlockState;
@@ -177,7 +179,14 @@
                 if (!tileentityshulkerbox.isCleared() && tileentityshulkerbox.shouldDrop())
                 {
                     ItemStack itemstack = new ItemStack(this);
-                    itemstack.getOrCreateTag().setTag("BlockEntityTag", ((TileEntityShulkerBox)tileentity).saveToNbt(new NBTTagCompound()));
+                    // [CM] Added nbttagcompound and if
+                    NBTTagCompound nbttagcompound = new NBTTagCompound();
+                    itemstack.getOrCreateTag().setTag("BlockEntityTag", ((TileEntityShulkerBox)tileentity).saveToNbt(nbttagcompound));
+                    if(!CarpetSettings.getBool("stackableEmptyShulkerBoxes") || !nbttagcompound.isEmpty())
+                    {
+                        itemstack.setTag(new NBTTagCompound());
+                    }
+                    // [CM] end
 
                     if (tileentityshulkerbox.hasCustomName())
                     {
