--- a/net/minecraft/block/BlockShulkerBox.java
+++ b/net/minecraft/block/BlockShulkerBox.java
@@ -2,6 +2,8 @@
 
 import java.util.List;
 import javax.annotation.Nullable;
+
+import carpet.CarpetSettings;
 import net.minecraft.block.material.EnumPushReaction;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.IBlockState;
@@ -177,7 +179,10 @@
                 if (!tileentityshulkerbox.isCleared() && tileentityshulkerbox.shouldDrop())
                 {
                     ItemStack itemstack = new ItemStack(this);
+                    if(!CarpetSettings.getBool("stackableShulkerBoxes") || !tileentityshulkerbox.isEmpty())
+                    { // really don't need to set nbt if its empty
                     itemstack.getOrCreateTag().setTag("BlockEntityTag", ((TileEntityShulkerBox)tileentity).saveToNbt(new NBTTagCompound()));
+                    }
 
                     if (tileentityshulkerbox.hasCustomName())
                     {
