--- a/net/minecraft/item/ItemPickaxe.java
+++ b/net/minecraft/item/ItemPickaxe.java
@@ -1,5 +1,6 @@
 package net.minecraft.item;
 
+import carpet.CarpetSettings;
 import com.google.common.collect.Sets;
 import java.util.Set;
 import net.minecraft.block.Block;
@@ -46,6 +47,7 @@
     public float getDestroySpeed(ItemStack stack, IBlockState state)
     {
         Material material = state.getMaterial();
-        return material != Material.IRON && material != Material.ANVIL && material != Material.ROCK ? super.getDestroySpeed(stack, state) : this.efficiency;
+        //[CM] Added Missing Tools For Piston And Glass
+        return material != Material.IRON && material != Material.ANVIL && material != Material.ROCK && (!CarpetSettings.getBool("missingTools") || (material != Material.PISTON && material != Material.GLASS)) ? super.getDestroySpeed(stack, state) : this.efficiency;
     }
 }
