--- a/net/minecraft/network/play/server/SPacketPlayerListHeaderFooter.java
+++ b/net/minecraft/network/play/server/SPacketPlayerListHeaderFooter.java
@@ -10,8 +10,8 @@
 
 public class SPacketPlayerListHeaderFooter implements Packet<INetHandlerPlayClient>
 {
-    private ITextComponent header;
-    private ITextComponent footer;
+    public ITextComponent header; // [CM] made public for HUD
+    public ITextComponent footer; // [CM] made public for HUD
 
     public void readPacketData(PacketBuffer buf) throws IOException
     {
